### ISSUE TITLE
Possible issue when using in parallel 

### ISSUE THREAD

**Issue by a6datta** [ssuluh/osrmnet#76]
@ssuluh I opened a new issue as this is different from NuGet package. Could you please check this sample and tell me what I am doing wrong when you have time. Thanks

```
public partial class Form1 : Form
    {        
        public Form1()
        {
            InitializeComponent();
        }

        private void button1_Click(object sender, EventArgs e)
        {
            GetDistanceAsync();
        }

        private void distanceTaskCallback(Task[] obj)
        {          
              this.BeginInvoke((Action)(()=>{
              foreach (Task<RouteResult> t in obj)
                      richTextBox1.Text =richTextBox1.Text+ Environment.NewLine + t.Result.Code;                    
              }));  
        }

        private RouteResult GetDistances(Coordinate start, Coordinate end) 
        {
            using (Osrm osrm = new Osrm(new EngineConfig() { StorageConfig = AppDomain.CurrentDomain.BaseDirectory + "great-britain-latest.osrm", Algorithm = Algorithm.CH, UseSharedMemory = false }))
            {
                RouteResult routeResult;
                var result = osrm.Route(new RouteParameters()
                {
                    Coordinates = new List<Coordinate>()
                    {
                        start,
                        end
                    },
                }, out routeResult);
                return routeResult;
            }            
        }
        private async Task GetDistanceAsync() 
        {
            List<Task> taskList = new List<Task>();
            Coordinate start = new Coordinate(51.559510, -0.132860);
            for (double ctr = 0.00001; ctr <= 0.00010; ctr=ctr+0.00001)
            {
                Coordinate end = new Coordinate(51.461280, -0.020340 + ctr);
                taskList.Add(Task.Factory.StartNew(() => GetDistances(start, end)));
            }            
            
            await Task.Factory.ContinueWhenAll(taskList.ToArray(), distanceTaskCallback);
        }
    }  
```

**Comment by a6datta** [comment]
It works this way. But it's not exactly a solution. Because if GetDistanceAsync() is called again before the previous set of tasks are completed, it will throw exception. Another option is to cancel the previous set of tasks, before starting new set. But it's also not ideal as two different methods may need to access the same data files. 

```
private RouteResult GetDistances(Coordinate start, Coordinate end, Osrm osrm) 
        {
            //using (Osrm osrm = new Osrm(new EngineConfig() { StorageConfig = AppDomain.CurrentDomain.BaseDirectory + "great-britain-latest.osrm", Algorithm = Algorithm.CH, UseSharedMemory = false }))
            //{
                RouteResult routeResult = new RouteResult();
                var result = osrm.Route(new RouteParameters()
                {
                    Coordinates = new List<Coordinate>()
                    {
                        start,
                        end
                    },
                }, out routeResult);
                Debug.WriteLine("Result: " + routeResult.Code);
                return routeResult;
            //}            
        }
        Osrm osrm;
        private async Task GetDistanceAsync() 
        {
            List<Task> taskList = new List<Task>();
            osrm = new Osrm(new EngineConfig() { StorageConfig = AppDomain.CurrentDomain.BaseDirectory + "great-britain-latest.osrm", Algorithm = Algorithm.CH, UseSharedMemory = false });
            Coordinate start = new Coordinate(51.559510, -0.132860);
            for (double ctr = 0.00001; ctr <= 0.00090; ctr = ctr + 0.00001)
            {
                Coordinate end = new Coordinate(51.461280, -0.020340 + ctr);
                taskList.Add(Task.Factory.StartNew(() => GetDistances(start, end, osrm)));
            }

            await Task.Factory.ContinueWhenAll(taskList.ToArray(), distanceTaskCallback);

            osrm.Dispose();
        }
```
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Exception thrown: 'Osrmnet.OsrmException' in osrm.net.dll
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Result: Ok
Exception thrown: 'Osrmnet.OsrmException' in osrm.net.dll
Result: Ok
Result: Ok
Result: Ok

**Comment by ssuluh** [comment]
@a6datta , can you help by trying this:  

Rather than walking the coordinate just send the same input coordinate in parallel (I don't cache anything on this side so it will always calculating again and again).  I want to see if you get the error due to osrm-backend throwing some exception because for some coordinate it fail to calculate distance somehow and throw exception or it is really parallel issue. 

Let me know, thanks

**Comment by a6datta** [comment]
@ssuluh 

I have tried this way. Here is the code and I have attached the error screenshot. 

```
public partial class Form1 : Form
    {        
        public Form1()
        {
            InitializeComponent();
        }

        private void button1_Click(object sender, EventArgs e)
        {
            GetDistanceAsync();
        }

        private void distanceTaskCallback(Task[] obj)
        {            
            foreach (Task<RouteResult> t in obj)
            {
                this.BeginInvoke((Action)(()=>{
                    if (t.Exception == null)
                        richTextBox1.Text = richTextBox1.Text+ Environment.NewLine +"Result1: "+ t.Result.Code+": "+t.Result.Routes[0].Distance.ToString();
                    else richTextBox1.Text = richTextBox1.Text + Environment.NewLine+"Error1: " + t.Exception.Message;
                }));                
            }            
        }
        private void distanceTaskCallback2(Task[] obj)
        {
            foreach (Task<RouteResult> t in obj)
            {
                this.BeginInvoke((Action)(() => {
                    if (t.Exception == null)
                        richTextBox1.Text = richTextBox1.Text + Environment.NewLine + "Result2: " + t.Result.Code + ": " + t.Result.Routes[0].Distance.ToString();
                    else richTextBox1.Text = richTextBox1.Text + Environment.NewLine + "Error2: " + t.Exception.Message;
                }));
            }
        }

        private RouteResult GetDistances(Coordinate start, Coordinate end) 
        {
            using (Osrm osrm = new Osrm(new EngineConfig() { StorageConfig = AppDomain.CurrentDomain.BaseDirectory + "great-britain-latest.osrm", Algorithm = Algorithm.CH, UseSharedMemory = false }))
            {
                RouteResult routeResult = new RouteResult();
                var result = osrm.Route(new RouteParameters()
                {
                    Coordinates = new List<Coordinate>()
                    {
                        start,
                        end
                    },
                }, out routeResult);
                Debug.WriteLine("Result: " + routeResult.Code);
                return routeResult;
            }            
        }        
        private async Task GetDistanceAsync() 
        {
            List<Task> taskList = new List<Task>();
            Coordinate start = new Coordinate(51.559510, -0.132860);
            for (double ctr = 0.00001; ctr <= 0.00090; ctr = ctr + 0.00001)
            {
                Coordinate end = new Coordinate(51.461280, -0.020340);
                taskList.Add(Task.Factory.StartNew(() => GetDistances(start, end)));
            }

            await Task.Factory.ContinueWhenAll(taskList.ToArray(), distanceTaskCallback);

        }
        private async Task GetDistanceAsync2()
        {
            List<Task> taskList = new List<Task>();
            Coordinate start = new Coordinate(51.559510, -0.132860);
            for (double ctr = 0.00001; ctr <= 0.00090; ctr = ctr + 0.00001)
            {
                Coordinate end = new Coordinate(51.461280, -0.020340);
                taskList.Add(Task.Factory.StartNew(() => GetDistances(start, end)));
            }

            await Task.Factory.ContinueWhenAll(taskList.ToArray(), distanceTaskCallback2);
        }

        private void button2_Click(object sender, EventArgs e)
        {
            GetDistanceAsync2();
        }
    }

**Comment by a6datta** [comment]
![error](https://user-images.githubusercontent.com/14138235/148408169-065d172d-3a23-4525-9290-af808687a77e.png)

**Comment by ssuluh** [comment]
It's been a while, but please confirm what I'm saying below by reading their documentation and search their issues. I think creating the actual EngineConfig itself is not threadsafe, but once an engine is created, you can call their service method from multiple thread.

If you look at your error it makes sense, you're trying to intialize the engine every time, during that it has to load the osrm files etc. and whatever else it needs to do.  I also know in Windows platform sometimes file being open and locked for no reason even after you free the pointer. If you look at osrmnet code in this project it just a thin wrapper, all the work is done by libosrm and we're not opening any file. Also the error clearly shows that exception comes from that side.

My recommendation to you is to try to create the Osrm object and the engine and store the reference to it. Then you can call the method from multiple threads.  Make sense?  Beside you don't want to do what you're doing there as creating the engine itself is an expensive operation both from memory and cpu perspective, you should be creating it one time and freeing them when you're done with them.

**Comment by a6datta** [comment]
I understand. I tried with a global config and it's working. Thank you
